src/pybind/mgr/volumes/fs/subvolume.py

```diff
@@ -143,6 +143,39 @@ def remove_subvolume(self, spec, force):
         except cephfs.Error as e:
             raise VolumeException(-e.args[0], e.args[1])
 
+    def resize_subvolume(self, subvolpath, newsize, noshrink):
+        """
+        :param subvolpath: subvolume path
+        :param newsize: new size In bytes
+        :return: new quota size and used bytes as a tuple
+        """
+        if newsize < 0:
+            raise VolumeException(-errno.EINVAL, "Provide a valid size")
+
+        try:
+            maxbytes = int(self.fs.getxattr(subvolpath, 'ceph.quota.max_bytes').decode('utf-8'))
+        except cephfs.NoData:
+            maxbytes = 0
+
+        subvolstat = self.fs.stat(subvolpath)
+        if newsize > 0 and newsize < subvolstat.st_size:
+            if noshrink:
+                raise VolumeException(-errno.EINVAL, "Can't resize the subvolume. The new size '{0}' would be lesser than the current "
+                                      "used size '{1}'".format(newsize, subvolstat.st_size))
+        elif newsize == 0:
+            if noshrink:
+                raise VolumeException(-errno.EINVAL, "Can't set quota to '{0}' when --no_shrink is specified. Remove --no_shrink and "
+                                      "try again.".format(newsize, subvolstat.st_size))
```

Comment:
Why should this error be raised? `... --no_shrink resize 0` should always succeed.

Reply:
@batrick It was originally written to fail when  `--no_shrink resize 0` is given. But then it seemed to be contradicting. `--no_shrink` means it shouldn't shrink below the current used size. Then why is it shrinking to size `0`, which is obviously below the current used size? It looked like `0` still got an entry (a bug) when `--no_shrink` was given.

I can remove that condition if you insist. What do you say?

Reply:
s/It was originally written to fail/It was originally written to succeed

Reply:
@batrick Yes, unset is better and we will continue disallowing `0` as invalid size. We had the unset discussion earlier here https://github.com/ceph/ceph/pull/30054#issuecomment-540913925.